libraries/platforms/dromajo-vcs/bsg_manycore_simulator.cpp

```diff
@@ -0,0 +1,662 @@
+      // before advancing to the next operation. This can prevent filling up of the
+      // FIFOs. However, not doing this can help in identifying situations that might
+      // create backpressure in actual hardware and provision for it.
+      do {
+        advance_time();
+        err = dpi->tx_req(*pkt, expect_response);
+      } while (err != BSG_NONSYNTH_DPI_SUCCESS  &&
+           (err == BSG_NONSYNTH_DPI_NO_CREDITS  ||
+            err == BSG_NONSYNTH_DPI_NO_CAPACITY ||
+            err == BSG_NONSYNTH_DPI_NOT_WINDOW  ||
+            err == BSG_NONSYNTH_DPI_BUSY        ||
+            err == BSG_NONSYNTH_DPI_NOT_READY));
+
+      if (err == BSG_NONSYNTH_DPI_SUCCESS) {
+        for (int i = 0;i < 4; i++)
+          host_to_mc_req_fifo->fifo[i].pop();
+      }
+      else {
+        bsg_pr_err("Packet transmission failed\n");
+        return HB_MC_FAIL;
+      }
+
+      // Update the credits in dromajo
+      err = dromajo_set_credits();
+      if (err != HB_MC_SUCCESS)
+        return err;
+    }
+    is_empty = mc_is_fifo_empty(type);
+  } while (!is_empty);
+
+  return HB_MC_SUCCESS;
+}
+
+/*
+ * dromajo_receive_packet
+ * Receives packets from the DPI FIFOs and pushes the data into the
+ * manycore request and response FIFOs in Dromajo
+ */
+int SimulationWrapper::dromajo_receive_packet() {
+  hb_mc_packet_t *mc_to_dromajo_req_packet, *mc_to_dromajo_resp_packet;
+  int err;
+  __m128i pkt;
+
+  // Read from the manycore request FIFO
+  // At every clock edge we are polling the request FIFO to see if there
+  // is a packet. If there is no packet (i.e err == BSG_NONSYNTH_NOT_VALID)
+  // we must move on
+  do {
+    do {
+      advance_time();
+      err = dpi->rx_req(pkt);
+    } while (err != BSG_NONSYNTH_DPI_SUCCESS    &&
+            (err == BSG_NONSYNTH_DPI_NOT_WINDOW ||
+             err == BSG_NONSYNTH_DPI_BUSY));
+
+    if (err == BSG_NONSYNTH_DPI_SUCCESS) {
+      mc_to_dromajo_req_packet = reinterpret_cast<hb_mc_packet_t *>(&pkt);
+      for (int i = 0; i < 4; i++) {
+        mc_to_host_req_fifo->fifo[i].push(mc_to_dromajo_req_packet->words[i]);
+      }
+    }
+  } while (err != BSG_NONSYNTH_DPI_NOT_VALID);
+
+  // Read from the manycore response FIFO
+  do {
+    do {
+      advance_time();
+      err = dpi->rx_rsp(pkt);
+    } while (err != BSG_NONSYNTH_DPI_SUCCESS    &&
+            (err == BSG_NONSYNTH_DPI_NOT_WINDOW ||
+             err == BSG_NONSYNTH_DPI_BUSY));
+
+    if (err == BSG_NONSYNTH_DPI_SUCCESS) {
+      mc_to_dromajo_resp_packet = reinterpret_cast<hb_mc_packet_t *>(&pkt);
+      for (int i = 0; i < 4; i++) {
+        mc_to_host_resp_fifo->fifo[i].push(mc_to_dromajo_resp_packet->words[i]);
+      }
+    }
+  } while (err != BSG_NONSYNTH_DPI_NOT_VALID);
+
+  return HB_MC_SUCCESS;
+}
+
+/*
+ * dromajo_set_credits
+ * Polls the hardware for credit information and sets the credits info
+ * for the Dromajo->Manycore request FIFO in dromajo
+ */
+int SimulationWrapper::dromajo_set_credits() {
```

Comment:
Change to update or poll, not set. Set implies something that has a argument, this changes an internal field.